src/EditorFeatures/Core/RenameTracking/RenameTrackingTaggerProvider.RenameTrackingCommitter.cs

```diff
@@ -120,48 +128,46 @@ private void ApplyChangesToWorkspace(CancellationToken cancellationToken)
+                    var result = TryUpdateWorkspaceForResetOfTypedIdentifier(workspace, renameTrackingSolutionSet.OriginalSolution, trackingSessionId);
+                    if (result is not null)
+                        return result;
+
+                    // Now that the solution is back in its original state, notify third parties about
+                    // the coming rename operation.
+                    if (!_refactorNotifyServices.TryOnBeforeGlobalSymbolRenamed(workspace, changedDocuments, renameTrackingSolutionSet.Symbol, newName, throwOnFailure: false))
+                        return EditorFeaturesResources.Rename_operation_was_cancelled_or_is_not_valid;
+
+                    // move all changes to final solution based on the workspace's current solution, since the current solution
+                    // got updated when we reset it above.
+                    var finalSolution = workspace.CurrentSolution;
+                    foreach (var docId in changedDocuments)
+                    {
+                        // because changes have already been made to the workspace (UpdateWorkspaceForResetOfTypedIdentifier() above),
+                        // these calls can't be cancelled and must be allowed to complete.
+                        var root = renameTrackingSolutionSet.RenamedSolution.GetDocument(docId).GetSyntaxRootSynchronously(CancellationToken.None);
+                        finalSolution = finalSolution.WithDocumentSyntaxRoot(docId, root);
+                    }
+
+                    // Undo/redo on this action must always clear the state machine
+                    return TryUpdateWorkspaceForGlobalIdentifierRename(
+                        workspace,
+                        finalSolution,
+                        _displayText,
+                        changedDocuments,
+                        renameTrackingSolutionSet.Symbol,
+                        newName,
+                        trackingSessionId);
                 }
-
-                // move all changes to final solution based on the workspace's current solution, since the current solution
-                // got updated when we reset it above.
-                var finalSolution = workspace.CurrentSolution;
-                foreach (var docId in changedDocuments)
+                finally
                 {
-                    // because changes have already been made to the workspace (UpdateWorkspaceForResetOfTypedIdentifier() above),
-                    // these calls can't be cancelled and must be allowed to complete.
-                    var root = renameTrackingSolutionSet.RenamedSolution.GetDocument(docId).GetSyntaxRootSynchronously(CancellationToken.None);
-                    finalSolution = finalSolution.WithDocumentSyntaxRoot(docId, root);
+                    RenameTrackingDismisser.DismissRenameTracking(workspace, changedDocuments);
                 }
```

Comment:
fairly certain not being in a finally here was a bug.  the return at https://github.com/dotnet/roslyn/pull/61515/files?diff=unified&w=1#diff-444ebec94ff45aa833894aec3cf66ed4c3f13f5f4fc661c5b5ce17d730597d48L140 meant we could bail out without dismissing our rename tracking state.